core/arch/mangle_shared.c

```diff
@@ -801,27 +847,59 @@ mangle_rseq(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr, instr_t *n
         }
 #    ifdef X86
         /* We just ran the instrumented version of the rseq code, with the stores
-         * removed.  Now we need to invoke it again natively for real.  We have to
-         * invoke the abort handler, not the rseq start, as it may perform some setup.
+         * removed.  Now we need to invoke it again natively for real.  We would prefer
+         * to invoke the abort handler, as it may perform some setup, but in too many
+         * cases it is truly an "abort" handler that just exits rather than a "restart
+         * handler".  Furthermore, to support executing a copy of the code natively in
+         * order to provide guarantees on regaining control and not rely on call-return
+         * semantics, it is simpler to execute only the limited-scope rseq region.
+         * Thus, we target the start point.
+         *
+         * In case the abort handler does perform setup, we checkpoint and restore GPR
```

Comment:
I forgot how do we ever find out if the abort handler was needed for setup?

Reply:
We don't find out.  We restore the original state just in case.

Reply:
And we document that we don't support any non-machine-state setup requirements.  See the entry in bt.dox:
```
- Each rseq region must handle being directly restarted without its
  abort handler being called (with the machine state restored).
```

Reply:
I was asking because the comment sounded like we somehow detect it "In case the abort handler does perform setup..." May want to clarify the comment.

Reply:
I think you're reading it as "In the case that the abort handler does perform setup".  Without "the" it means "Just in case" or "On the chance that".

Reply:
Ok :)